compiler/rustc_trait_selection/src/traits/specialize/mod.rs

```diff
@@ -178,101 +265,117 @@ pub(super) fn specializes(tcx: TyCtxt<'_>, (impl1_def_id, impl2_def_id): (DefId,
-
-    let selcx = SelectionContext::new(infcx);
-    let target_args = infcx.fresh_args_for_item(DUMMY_SP, target_impl);
-    let (target_trait, obligations) =
-        util::impl_subject_and_oblig(&selcx, param_env, target_impl, target_args, error_cause);
+    let parent_args = infcx.fresh_args_for_item(DUMMY_SP, parent_impl_def_id);
+    let parent_impl_trait_ref = ocx.normalize(
+        cause,
+        param_env,
+        infcx
+            .tcx
+            .impl_trait_ref(parent_impl_def_id)
+            .expect("expected source impl to be a trait impl")
+            .instantiate(infcx.tcx, parent_args),
+    );
 
     // do the impls unify? If not, no specialization.
-    let Ok(InferOk { obligations: more_obligations, .. }) = infcx
-        .at(&ObligationCause::dummy(), param_env)
-        // Ok to use `Yes`, as all the generic params are already replaced by inference variables,
-        // which will match the opaque type no matter if it is defining or not.
-        // Any concrete type that would match the opaque would already be handled by coherence rules,
-        // and thus either be ok to match here and already have errored, or it won't match, in which
-        // case there is no issue anyway.
-        .eq(DefineOpaqueTypes::Yes, source_trait, target_trait)
+    let Ok(()) = ocx.eq(cause, param_env, specializing_impl_trait_ref, parent_impl_trait_ref)
     else {
-        debug!("fulfill_implication: {:?} does not unify with {:?}", source_trait, target_trait);
-        return Err(());
+        return false;
     };
 
-    // attempt to prove all of the predicates for impl2 given those for impl1
-    // (which are packed up in penv)
-    ocx.register_obligations(obligations.chain(more_obligations));
+    // Now check that the source trait ref satisfies all the where clauses of the target impl.
+    // This is not just for correctness; we also need this to constrain any params that may
+    // only be referenced via projection predicates.
+    let predicates = ocx.normalize(
+        cause,
+        param_env,
+        infcx.tcx.predicates_of(parent_impl_def_id).instantiate(infcx.tcx, parent_args),
+    );
+    let obligations = predicates_for_generics(|_, _| cause.clone(), param_env, predicates);
+    ocx.register_obligations(obligations);
 
     let errors = ocx.select_all_or_error();
     if !errors.is_empty() {
         // no dice!
         debug!(
             "fulfill_implication: for impls on {:?} and {:?}, \
                  could not fulfill: {:?} given {:?}",
-            source_trait,
-            target_trait,
+            specializing_impl_trait_ref,
+            parent_impl_trait_ref,
             errors,
             param_env.caller_bounds()
         );
-        return Err(());
+        return false;
     }
 
-    debug!("fulfill_implication: an impl for {:?} specializes {:?}", source_trait, target_trait);
+    // If the parent impl is const, then the specializing impl must be const.
```

Comment:
Note to self: Actually I don't know if this is correct -- well, it's not complete.

```
const trait Bar {}
impl const Bar for T {}

const trait Foo {}
impl const Foo for T where T: ~const Bar {}
// specializing impl:
impl Foo for (T,) {}
```

The specializing impl satisfies `(T,): ~const Bar` because of the blanket impl, but it's still less const than the parent impl. I think we need to check that the specializing impl is *also* conditionally const, and then bail.